src/python/grpcio/grpc/_interceptor.py

```diff
@@ -133,23 +185,35 @@ def __init__(self, thunk, method, interceptor):
         self._interceptor = interceptor
 
     def __call__(self, request, timeout=None, metadata=None, credentials=None):
-        call_future = self.future(
+        response, ignored_call = self.with_call(
             request,
             timeout=timeout,
             metadata=metadata,
             credentials=credentials)
-        return call_future.result()
+        return response
 
     def with_call(self, request, timeout=None, metadata=None, credentials=None):
-        call_future = self.future(
-            request,
-            timeout=timeout,
-            metadata=metadata,
-            credentials=credentials)
-        return call_future.result(), call_future
+        client_call_details = _ClientCallDetails(self._method, timeout,
+                                                 metadata, credentials)
 
-    def future(self, request, timeout=None, metadata=None, credentials=None):
+        def continuation(new_details, request):
+            new_method, new_timeout, new_metadata, new_credentials = (
+                _unwrap_client_call_details(new_details, client_call_details))
+            try:
+                response, call = self._thunk(new_method).with_call(
+                    request,
+                    timeout=new_timeout,
+                    metadata=new_metadata,
+                    credentials=new_credentials)
+                return _UnaryOutcome(response, call)
+            except Exception as exception:  # pylint:disable=broad-except
```

Comment:
If I am understanding things correctly, I think this might incorrectly swallow RpcError codes. `with_call` raises on non-OK RPC termination and then _FailureOutcome sets the `code()` to internal, rather than the original RpcCode.

The approach I went with is catching `RpcError` (which when raised from `with_call` is actually a _Rendezvous/Call/Future) and returning that. 

In my tests this keeps the same API for interceptors, that `continuation(call_details,request)` returns a Call/Future which they can inspect/replace and if the interceptor does nothing but call the continuation then its the same as if there is no interceptor. 


```
def install_monkeypatches():
    import sys
    import grpc._interceptor
    from grpc._interceptor import _ClientCallDetails, _LocalFailure

    def blocking_unary_with_call(self, request, timeout=None, metadata=None, credentials=None):
        def continuation(client_call_details, request):
            channel = self._thunk(client_call_details.method)
            try:
                resp, call = channel.with_call(request,
                                               timeout=client_call_details.timeout,
                                               metadata=client_call_details.metadata,
                                               credentials=client_call_details.credentials)
                return call
            except grpc.RpcError as e:
                return e

        client_call_details = _ClientCallDetails(self._method, timeout,
                                                 metadata, credentials)
        try:
            resp = self._interceptor.intercept_unary_unary(
                continuation, client_call_details, request)
        except Exception as exception:  # pylint:disable=broad-except
            resp = _LocalFailure(exception, sys.exc_info()[2])
        return resp.result(), resp

    grpc._interceptor._UnaryUnaryMultiCallable.with_call = blocking_unary_with_call

    def blocking_unary_call(self, request, timeout=None, metadata=None, credentials=None):
        resp, call = self.with_call(request,
                                    timeout=timeout,
                                    metadata=metadata,
                                    credentials=credentials)
        return resp

    grpc._interceptor._UnaryUnaryMultiCallable.__call__ = blocking_unary_call
```

Reply:
You are correct; special casing `RpcError` before the board-exception-handler will fix that situation. I will add a test to handle that one before fixing it.